stew/base36.nim

```diff
@@ -0,0 +1,222 @@
+
+{.push raises: [].}
+
+type
+  Base36Status* {.pure.} = enum
+    Error,
+    Success,
+    Incorrect,
+    Overrun
+
+  Base36Alphabet* = object
+    decode*: array[128, int8]
+    encode*: array[36, uint8]
+
+  Base36* = object
+  Base36Uc* = object
+  Base36Lc* = object
+
+  Base36Types* = Base36 | Base36Uc | Base36Lc
+
+  Base36Error* = object of CatchableError
+    ## Base36 specific exception type
+
+func newAlphabet36(s: string): Base36Alphabet =
+  doAssert(len(s) == 36)
+  var alphabet: Base36Alphabet
+  for i in 0..<len(s):
+    alphabet.encode[i] = cast[uint8](s[i])
+  for i in 0..<len(alphabet.decode):
+    alphabet.decode[i] = -1
+  for i in 0..<len(alphabet.encode):
+    alphabet.decode[int(alphabet.encode[i])] = int8(i)
+  return alphabet
+
+const
+  B36UcAlphabet* = newAlphabet36("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ")
+  B36LcAlphabet* = newAlphabet36("0123456789abcdefghijklmnopqrstuvwxyz")
+
+func encodedLength*(btype: typedesc[Base36Types], length: int): int =
+  ## Return estimated length of Base36 encoded value for plain length
+  ## ``length``.
+  return (length * 277) div 179 + 1
+
+func decodedLength*(btype: typedesc[Base36Types], length: int): int =
+  ## Return estimated length of decoded value of Base36 encoded value of length
+  ## ``length``.
+  return (length * 179) div 277
+
+func encode*(btype: typedesc[Base36Types], inbytes: openArray[byte],
+             outstr: var openArray[char], outlen: var int): Base36Status =
+  when (btype is Base36) or (btype is Base36Lc):
+    const alphabet = B36LcAlphabet
+  elif (btype is Base36Uc):
+    const alphabet = B36UcAlphabet
+
+  let binsz = inbytes.len
+  var zcount = 0
+  while zcount < binsz and inbytes[zcount] == 0x00'u8:
+    inc(zcount)
+
+  let size = ((binsz - zcount) * 277) div 179 + 1
+  var buffer = newSeq[uint8](size)
+
+  var hi = size - 1
+  for i in zcount..<binsz:
+    var carry = uint32(inbytes[i])
+    var j = size - 1
+    while (j > hi) or (carry != 0'u32):
+      carry = carry + uint32(256'u32 * buffer[j])
+      buffer[j] = cast[byte](carry mod 36)
+      carry = carry div 36
+      dec(j)
+    hi = j
+
+  # Strip leading zeros in encoded buffer
+  var j = 0
+  while j < size and buffer[j] == 0x00'u8:
+    inc(j)
+
+  outlen = 1 + zcount + (size - j) # 1 for 'k'
+  if outstr.len < outlen:
+    return Base36Status.Overrun
+
+  outstr[0] = 'k'
+  for k in 0..<zcount:
+    outstr[1 + k] = cast[char](alphabet.encode[0])
+
+  var i = 1 + zcount
+  while j < size:
+    outstr[i] = cast[char](alphabet.encode[buffer[j]])
+    inc(j)
+    inc(i)
+
+  return Base36Status.Success
+
+
+func encode*(btype: typedesc[Base36Types],
+             inbytes: openArray[byte]): string {.inline.} =
+  ## Encode array of bytes ``inbytes`` using Base36 encoding and return
+  ## encoded string.
+  var size = (len(inbytes) * 277) div 179 + 1 + 1
```

Comment:
Is the additional `+ 1` over `encodedLength` a maximum value? This in any case could also use `encodedLength`, since right now basically the same calculation is written in 3 different places.

Reply:
Or because of the `k`?